### ISSUE TITLE
Support Markdown dialect syntax for math/formula blocks

### ISSUE THREAD

**Issue by arwedus** [doxygen/doxygen#8009]
**Expected behavior**
Currently, latex formulas in inline spans or blocks can be used in Markdown files to be parsed by Doxygen, using the Doxygen commands \f$ and \f[ \f], respectively.

There are some Markdown implementations out there which generate LaTeX or MathJax output when they encounter a formula. Unfortunately, they could not align on a syntax (ideally, they would have chosen the clean and unambiguous Doxygen syntax, of course).

Since the feature is so non-standardized, I understand why it has not made it into the Doxygen Markdown parser. However, the "Jupyter Notebook" variant seems to the most popular among those I found (e.g. https://marketplace.visualstudio.com/items?itemName=goessner.mdmath ):

  - Single $ signs delimit an inline $math section $ (must be without whitespace between the first $ sign and the math section).
  - Double $$ signs delimit a $$ math block $$ (not sure about the spacing there).

Other styles are e.g. \(inline formula\) and \[ block formula \] or \$$ inline formula $$ and $$ block formula $$ (kramdown).

An option would be to add a config file option for Markdown math command parsing. (i.e. that the user enters the desired delimiter style).

I already wrote myself an input filter that naively transforms the Jupyter Notebook style into Doxygen style using sed. However, this does not respect any non-translated sections (i.e. code), so it's not a very good solution.

**To Reproduce**

Put the text inside the code block into a md file and include it to test for the "Jupyter notebook" / Dollar style.

```
$\alpha$ to be replaced, `$\gamma$` and $ delta $ not. $10 also not.

first test $$ equation = \alpha $$.

 \$alpha$ not replaced, \f$x^y\f$ to be preserved, $\beta = x^y\$ to be replaced.

Second test: 
$$
\left[ \alpha \right]
$$
end.
```

**Version**
1.8.20

**Comment by albert-github** [comment]
Does Jupyter have anywhere written down the markdown the syntax they use?
Why should `$ delta $` not be replaced it is just the text delta.

**Comment by arwedus** [comment]
They have, in two sentences: https://jupyter-notebook.readthedocs.io/en/stable/examples/Notebook/Working%20With%20Markdown%20Cells.html

Their actual implementation is as primitive as the spec: Whatever is between the first and the next Dollar sign on a line will be passed to MathJax. So let's say you write:

    I owe you $10. This is a formula: $\int_0^\infty x^{-\alpha}$ 

Then the text between $10 and then next $ sign will be parsed. To escape $ signs in Jupyter notebook, one needs to write \\$.

This is actually the problem with single $ signs in text, they could appear in other context. Okay... maybe another Markdown style would be a better choice.

**Comment by albert-github** [comment]
Not a very convincing and good description by Jupyter.
3 $ signs like in your example always will be a problem as one doesn't directly know which one is a formula starter / ending and which is a dollar sign. A dollar sign it not only used for amounts of money, but also for *nix environment variables, php variable names and there are probably there a some more examples.

**Comment by arwedus** [comment]
True. It's not really their use case though. Anyhow, since this is not generic enough I'd rather try to convince that VS code plugin developer to add the Doxygen syntax, which is generic and unambiguous, than the other way around.

**Comment by arwedus** [comment]
The style rules I mentioned in the original post is actually pandoc (https://pandoc.org/MANUAL.html#math ). This is less ambiguos but not perfect either.

The kramdown/pandoc/paperhive syntax is much less ambiguous (https://kramdown.gettalong.org/syntax.html#math-blocks and ). It basically uses double $$ for inline (span) as well as block math. So $$ this is inline math $$ and

$$ this is block math $$

and if you want a formula that begins at a newline to be inlined, you have to escape the first $ sign
\$$ like this $$.

Maybe the kramdown math style is worth supporting in a future version?

**Comment by doxygen** [comment]
@arwedus I suggest to request a 'doxygen' style being added here: https://github.com/goessner/markdown-it-texmath
If I look at the `texmath.rules` array in the `texmath.js` file it should be pretty easy.

**Comment by arwedus** [comment]
@doxygen  I think so too, but that will only serve the few people that use that VS code plug-in (around 100.000 judging from the download count).

[Pandoc style](https://pandoc.org/MANUAL.html#math) is actually just a more precise specification of the original TeX specification: $...$ is math, $$...$$ is displaymath. (<https://texblog.net/help/latex/ltx-421.html>).

In addition, many Markdown parsers support:

    ```math
    \sum a_i
    ```

which I think is a sensible

Since I've opened this issue, someone else create a comprehensive overview of all markdown-math-delimiters existing in the field at <https://github.com/cben/mathdown/wiki/math-in-markdown>. Scroll down to the bottom to find an overview table... it shows that almost every parser supporting Latex math supports $...$ and $$...$$ delimiters.

I would use this as the spec for such a parser (copied from pandoc and extended a little bit):

* Anything between two $ characters will be treated as TeX math. The opening $ must have a non-space character immediately to its right, while the closing $ must have a non-space character immediately to its left, and must not be followed immediately by a digit.
* If you need to enclose text in literal $ characters, backslash-escape them and they won’t be treated as math delimiters.
* For display math, use $$ delimiters. (In this case, the delimiters may be separated from the formula by whitespace)
* + Opening and closing $ must be on the same line (or, slightly more permissive: In the same paragraph).
* + $ delimiters are greedy: The formula `$\alpha =$(a+b)$` ends at `)b`.
* + extended: Delimiters inside code/verbatim blocks are not treated as math delimiters.

Would you agree with that spec for a feature request?

As with most parsers, the behavior could be enabled by a config option (e.g. FORMULA_DOLLARMATH).

**Comment by albert-github** [comment]
I think that when this syntax is properly defined in an independent standard it is worthwhile to think about adding it to doxygen as well. At the moment I'm not aware of any independent standard that supports it.

An interesting idea in this respect is though https://github.com/commonmark/commonmark-spec/issues/694

**Comment by nilsvu** [comment]
A great overview is compiled here: https://github.com/cben/mathdown/wiki/math-in-markdown#appendix-summary-table

Based on that overview I think support for $...$, $$...$$ and naked \begin{equation}...\end{equation} is the de-facto standard, supported by the majority of parsers with slight variations on the specific implementation.

**Comment by arwedus** [comment]
@nilsvu , thanks for sharing the overview, I used that as a reference as well when I created this feature request.

I'll open source my doller-math input filter as soon as I can find the time 🙂

**Comment by nilsvu** [comment]
FWIW, here's the Perl filter I ended up implementing to convert from markdown to Doxygen-style math: https://github.com/sxs-collaboration/spectre/blob/develop/docs/config/markdown-math-filter.pl

**Comment by aprsa** [comment]
In light of github's MD support, I'd like to bump this suggestion. Pre-processing can of course always be used as a workaround, but I agree with @nilsvu that support for `$ ... $` and `$$ ... $$`, and possibly `\begin{equation} ... \end{equation}` has become the de-facto standard. It would be very useful to have doxygen's base MD be consistent with github's MD flavor.

**Comment by debemdeboas** [comment]
Bumping this as well.
Doxygen's own documentation mentions using GitHub READMEs as a main page (see `USE_MDFILE_AS_MAINPAGE`), so supporting GitHub's MD or CommonMark should be a given.
